src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c

```diff
@@ -151,55 +151,107 @@ static inline void Log_Debug_With_Frequency(int freqSeconds, const char* format,
-    if (!interface)
+    if (interface == NULL)
+    {
         return NULL;
-    ADU_MQTT_COMMUNICATION_MGR_STATE* state = (ADU_MQTT_COMMUNICATION_MGR_STATE*)interface->moduleData;
-    return state;
+    }
+
+    return (ADU_MQTT_COMMUNICATION_MGR_STATE*)interface->moduleData;
 }
 
 /**
- * @brief Create the communication channel module.
+ * @brief Create and Init Communication Mgr State object
+ *
+ * @return ADU_MQTT_COMMUNICATION_MGR_STATE* The new communication manager state instance, or NULL on failure.
+ * @remark Caller is responsible for calling DestroyCommunicationMgrState on the returned state instance.
  */
-ADUC_AGENT_MODULE_HANDLE ADUC_Communication_Channel_Create()
+static ADU_MQTT_COMMUNICATION_MGR_STATE* CreateCommunicationMgrState()
 {
-    ADUC_AGENT_MODULE_INTERFACE* interface = NULL;
     ADU_MQTT_COMMUNICATION_MGR_STATE* state = NULL;
 
-    ADUC_ALLOC(state);
+    ADU_MQTT_COMMUNICATION_MGR_STATE* tmp = calloc(1, sizeof(*tmp));
+    if (tmp == NULL)
+    {
+        goto done;
+    }
 
-    state->pendingSubscriptions = VECTOR_create(sizeof(ADUC_MQTT_SUBSCRIBE_CALLBACK_INFO));
-    if (state->pendingSubscriptions == NULL)
+    tmp->pendingSubscriptions = VECTOR_create(sizeof(ADUC_MQTT_SUBSCRIBE_CALLBACK_INFO));
+    if (tmp->pendingSubscriptions == NULL)
     {
-        Log_Error("Failed to create pending subscriptions vector");
         goto done;
     }
 
-    Log_Debug("pendingSubscriptions obj created @ %p", state->pendingSubscriptions);
+    tmp->connectionRetryParams = defaultConnectionRetryParams;
+
+    state = tmp;
+    tmp = NULL;
+
+done:
+    free(tmp);
+
+    return state;
+}
+
+/**
+ * @brief Destroys the fields of communication manager state instance.
+ *
+ * @param state The state.
+ */
+static void DestroyCommunicationMgrState(ADU_MQTT_COMMUNICATION_MGR_STATE* state)
+{
+    if (state != NULL)
+    {
+        VECTOR_destroy(state->pendingSubscriptions);
+        state->pendingSubscriptions = NULL;
+
+        if (state->subscribeTopicInfo.topic != NULL)
+        {
+            free(state->subscribeTopicInfo.topic);
+            memset(&state->subscribeTopicInfo, 0, sizeof(state->subscribeTopicInfo));
+        }
+    }
+}
+
+/**
+ * @brief Create the communication channel module.
+ * @returns ADUC_AGENT_MODULE_HANDLE The created module handle.
+ * @remark Caller is responsible for calling ADUC_Communication_Channel_Destroy on the handle.
+ */
+ADUC_AGENT_MODULE_HANDLE ADUC_Communication_Channel_Create()
+{
+    ADUC_AGENT_MODULE_HANDLE handle = NULL;
+    ADU_MQTT_COMMUNICATION_MGR_STATE* state = NULL;
+    ADUC_AGENT_MODULE_INTERFACE* interface = NULL;
 
-    memset(&state->subscribeTopicInfo, 0, sizeof(state->subscribeTopicInfo));
-    state->connectionRetryParams = defaultConnectionRetryParams;
+    state = CreateCommunicationMgrState();
+    if (state == NULL)
+    {
+        goto done;
+    }
 
-    ADUC_ALLOC(interface);
+    interface = calloc(1, sizeof(*interface));
```

Comment:
why is interface a vector of size 1?  In the future will we have multiple interfaces at the same time, and if so, how does ADUC_AGENT_MODULE_HANDLE know the size of ".interface" ?

Reply:
maybe this is being used for zero-init?

Reply:
yes, using for zero-init.